### ISSUE TITLE
No sane way to handle closed InputStream with Observable.fromInputStream

### ISSUE THREAD

**Issue by mdedetrich** [monix/monix#1102]
I am attempting to use `Observable` to handle `InputStream`'s which are the result of `java.lang.Process`. Java's `Process` exposes two streams, the `STDOUT` stream and the `STDERR` stream. If the process is running successfully then the `InputStream` for `STDERR` will already be closed (and vice versa)

The problem is that appears Monix Observable is unable to properly handle a closed stream. When you try to read from a closed stream that is already closed you get a `java.io.IOException: Stream closed` however there is no way to catch this error. Functions such as `onErrorRecover` or `onErrorHandle` (or their variants) don't work.

The intention is to do something like this

```scala
val inputStreamObservable =
  Observable
    .fromInputStream(processTask.map(_.getInputStream))
    .map(bytes => Right(new String(bytes)): Either[String, String]).onErrorRecoverWith {
    case _: IOException => Observable.never
  }

val errorStreamObservable =
  Observable
    .fromInputStream(processTask.map(_.getErrorStream))
    .map(bytes => Left(new String(bytes)): Either[String, String]).onErrorRecoverWith {
    case _: IOException => Observable.never
  }

Observable
  // We want to get either the error if it exists on STDERR otherwise we just want STDIN
  .firstStartedOf(inputStreamObservable, errorStreamObservable)
```

Basically I want to have an `Observable` which holds an `Either`, if its a `Left` then we have a `STDERR` and if its right then we have `STDOUT`

**Comment by jvican** [comment]
I haven't implemented something like this myself, but if I was I would use [nuprocess]() instead of `java.lang.Process` because it heavily simplifies the handling of process output streams. You can just implement your own `NuAbstractProcessHandler` and have `onStdout` and `onStderr` push the contents to your own observer and then consume it with observable.

**Comment by mdedetrich** [comment]
I am having issues replicating this outside of the context of work project, so I need to investigate it.

@jvican Thanks for the tip, will look into https://github.com/brettwooldridge/NuProcess

**Comment by mdedetrich** [comment]
I'm going to close this since I can't sanely reproduce, will reopen if I discover something new
